### PR TITLE
Refactor module into a class, add custom errors

### LINKED ISSUE #1: Extend expected errors of jwt-client
On service side we need to provide proper response with 400/401 status code if token doesn't have valid payload structure or has missing fields. But current errors are generic and cannot be differentiated from other runtime errors.

As a solution you can add some 'marker' or flag to error objects.
Also you can move checks from getter functions to 'jwtClient' function. In such case we will wrap creation of jwtClient in try-catch and handle errors there.

### PR DESCRIPTION
**Pull request by EthanJStark**
Closes #1

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,60 +1,96 @@
...
+      tenants: this._payload.tenants
+    })
   }
 }
 
-module.exports = jwtClient
+const makeJwtClient = jwt => new JwtClient(jwt)
+makeJwtClient.Error = JwtError
+makeJwtClient.isAJwtError = JwtError.isAJwtError
+module.exports = makeJwtClient
```
Comment (neezer) [inline index.js]: Uh, nope. `isAJwtError` should be provided as a separate export, not as a static method on `makeJwtClient`.

```js
exports.isAJwtError = JwtError.isAJwtError

exports['default'] = jwt => new JwtClient(jwt)

module.exports = Object.assign(exports['default'], exports)
```

File: index.js [hunk 2]
```diff
@@ -1,60 +1,96 @@
...
-    ort: getORT(),
-    tenants: payload.tenants
-  })
+  get getTenantId () {
+    return () => this._tenantId
+  }
+
+  get getORT () {
+    return () => this._ort
+  }
```
Comment (neezer) [inline index.js]: `getUserId`, `getTenantId`, and `getORT` should all be methods, not getters.

For example, 

```diff
- get getOrt () {
-   return () => this._ort
- }

+ getORT () {
+   return this._ort
+ }
```

File: index.js [hunk 3]
```diff
@@ -1,60 +1,96 @@
...
-    getORT,
-    getRoleIdForTenant,
-    dump
+  get dump () {
+    return () => ({
+      id: this._userId,
+      tenantId: this._tenantId,
+      ort: this._ort,
+      tenants: this._payload.tenants
+    })
```
Comment (neezer) [inline index.js]: Same as above; implement as a method, not a getter.

File: index.js [hunk 4]
```diff
@@ -1,60 +1,96 @@
...
-  if (!payload) {
-    throw new Error('Invalid JWT!')
+class JwtClient {
+  constructor (jwt) {
+    this._jwt = this._split(jwt)
+    this._payload = this._assertShape(this._decode(this._jwt))
+    this._userId = this._payload.id
+    this._tenantId = this._payload.tenantId
+    this._ort = this._payload.ort
   }
```
Comment (neezer) [inline index.js]: I'd like to see the shape assertions be separate from the payload assignment, something like

```js
constructor (jwt) {
  // the innards of _decode should deal with "Bearer "
  this._payload = this._decode(jwt)
  
  if (this._payloadLooksFunny()) {
    throw new JwtError(/* ... */)
  }

  this._ort = this._payload.ort
  // etc...
}
```

File: index.js [hunk 5]
```diff
@@ -1,60 +1,96 @@
...
+      }
+    }
+    return jwt
+  }
 
-  const getRoleIdForTenant = tenantId => {
-    if (!payload.tenants.hasOwnProperty(tenantId)) {
-      throw new Error(`No roleId for tenant ${tenantId} in JWT Payload`)
+  _decode (jwt) {
+    const payload = jsonwebtoken.decode(jwt)
```
Comment (neezer) [inline index.js]: This should be wrapped. What if you try to parse a `jwt === "="`? `decode` will throw but it won't throw our custom error, so any `catch` written by the consumer has to account for two types of errors: a `JwtError` and whatever the above monkey wrench throws.

Comment (neezer) [reply index.js]: Nm, it appears `jsonwebtoken.decode` doesn't throw errors when given invalid values.

File: index.js [hunk 6]
```diff
@@ -80,17 +63,18 @@ class JwtClient {
     return this._payload.tenants[tenantId].authorization
```
Comment (neezer) [inline index.js]: What happens when ... ?

```js
this._payload.tenants[tenantId] === undefined
```

Comment (EthanJStark) [reply index.js]: Good catch!

**Comment by EthanJStark** [comment]
@neezer Change addressed, should be good

File: index.js [hunk 7]
```diff
@@ -1,60 +1,84 @@
 const jsonwebtoken = require('jsonwebtoken')
+const R = require('ramda')
```
Comment (neezer) [inline index.js]: The previous rejection message pertains to this line. Sorry, pasted my response in the wrong box. 🙊

Comment (EthanJStark) [reply index.js]: Good call; shorter `R.contains` syntax definitely not worth bringing `ramda` into `dependencies`.

**Comment by EthanJStark** [comment]
@neezer Change addressed!

**Pull request merged**